RFCs/FS-1092-map-min-max-keyvalue.md

```diff
@@ -0,0 +1,87 @@
+# F# RFC FS-1092- Add Map.maxKeyValue and Map.minKeyValue functions
+
+The design suggestion [Add Map.maxBinding and Map.minBinding functions](https://github.com/fsharp/fslang-suggestions/issues/933) has been marked "approved in principle".
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] Approved in principle
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/933)
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/FILL-ME-IN)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion](https://github.com/fsharp/fslang-design/issues/PLEASE-ADD-A-DISCUSSION-ISSUE-AND-LINK-HERE)
+
+# Summary
+
+This feature extends the `Map` module by adding two functions, tentatively named
+`Map.maxKeyValue` and `Map.minKeyValue`. Those functions allow users to look up
+the current minimum/maximum key-value pair from a map without having to iterate
+through the whole key-value pairs.
+
+# Motivation
+
+Since the `Map` represents an *ordered* map, users often expect to find or use
+the largest/smallest key from the map. Consider the following example where
+there is a hypothetical database implementation that stores age-name mappings.
+
+```fsharp
+let db: Map<uint32, string list> = Map.empty
+
+let addMember age name db =
+  match Map.tryFind age db with
+  | Some names -> Map.add age (name :: names) db
+  | None -> Map.add age [ name ] db
+
+let findLargestAge db =
+  (* Hypothetical example usage of maxKeyValue *)
+  Map.maxKeyValue db |> fst
+
+let main () =
+  (* Add members to the db *)
+  let db = addMember 20 "Alice"
+  let db = addMember 20 "Bob"
+  let db = addMember 42 "Charlie"
+  (* Find the largest age in the db *)
+  findLargestAge db |> printfn "%d"
+```
+
+In this case, we would like to find out the largest age in the
+database. Currently the only way to get the largest age is to use `Map.filter`
+or `Map.fold`, which requires iterating the whole entries in the worst case.
+
+# Detailed design
+
+I didn't look at the actual implementation of `Map` yet, but it should be
```

Comment:
For the purposes of the document, I recommend looking into the implementation and then writing it from the perspective of:

"The implementation is `O(log n)` by doing [...]"

Or something to that effect. Since maps are currently backed by an AVL tree that would be true.

Reply:
It was true. I modified the description accordingly.